### ISSUE TITLE
InsertAndFetch with MySQL 8,  UUID() function and default Id, return id 0

### ISSUE THREAD

**Issue by pavellob** [Vincit/objection.js#1706]
I have to using MySQL in my work, version 8.0.13
As id I used UUID, wich generated by internal mySQL function UUID() // it possible from MySQL 8
my knex migration  contains something like ```table.uuid('id').primary().defaultTo(knex.raw('(UUID())'));```

And for seeds it all work fine.
When I use insertAndFetch, row created with valid uuid, but return the object with **id:0**
something like 
```
UserModel {
  roleId: 'd251f5d8-5fbe-11ea-ab31-0242ac110002',
  email: 'test2@gmail.com',
  timezone: 'US/Mountain',
  id: 0, // real id is 4a11a33c-5fbf-11ea-ab31-0242ac110002
  role: RoleModel {
    id: 'd251f5d8-5fbe-11ea-ab31-0242ac110002',
    name: 'mediaplan creator',
    access: [ [Object] ],
    createdAt: 2020-03-06T12:26:16.000Z,
    updatedAt: 2020-03-06T12:26:16.000Z,
    deletedAt: null
  }
}
```

As I understand, it should return my UUID, but return 0

**Comment by koskimas** [comment]
I can't reproduce this. I'll need a reproduction script from you.

**Comment by dsul** [comment]
Hi @pavellob. I had a very similar scenario as the one you described here (same database and version, same migration code, same result). I will share my thoughts and resolution in the hope that it helps you or helps reveal the underlying issue as a whole. 

_I am not intimately familiar with the database technologies so my theory on the issue might be a coincidence admittedly._

MySQL uses a `LAST_INSERT_ID` function that states
> With no argument, LAST_INSERT_ID() returns a BIGINT UNSIGNED (64-bit) value representing the first automatically generated value successfully inserted for an AUTO_INCREMENT column as a result of the most recently executed INSERT statement.

https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id

The keyword here is `AUTO_INCREMENT`. Since we modified our primary id field to be of type `char(36)`, this function will not give the value we're expecting. 

Similarly, the C API function that deals with a similar value on INSERTS and UPDATES says
> The return value of mysql_insert_id() is always zero unless explicitly updated under one of the following conditions:

https://dev.mysql.com/doc/refman/8.0/en/mysql-insert-id.html

None of the following conditions matched for myself.


So, it seems like it might be possible that the `andFetch` functions, when relying on the database to generate the ID, would run one of the above functions and return zero by default due to no longer having the appropriate field type. 

The solution I chose was to generate the UUID in the JS code. Using a `$beforeInsert` model instance hook, I set the model's ID to the generated UUID. Relying on the JS to generate and set the UUID over the database seems to bypass the need for the database to call the above functions (or at least, not use their output) since the ID is already available to it, which it can then just correctly return.
